spring-integration-core/src/main/java/org/springframework/integration/dispatcher/PartitionedDispatcher.java

```diff
@@ -0,0 +1,191 @@
+ * the message to dispatch.
+ * <p>
+ * Every partition, created by this class, is a {@link UnicastingDispatcher}
+ * delegate based on a single thread {@link Executor}.
+ * <p>
+ * The number of partitions should be a reasonable value for the application environment
+ * since every partition is based on a dedicated thread for message processing.
+ * <p>
+ * The rest of the logic is similar to {@link UnicastingDispatcher} behavior.
+ *
+ * @author Artem Bilan
+ *
+ * @since 6.1
+ */
+public class PartitionedDispatcher extends AbstractDispatcher {
+
+	private final Map<Integer, UnicastingDispatcher> partitions = new HashMap<>();
+
+	private final List<ExecutorService> executors = new ArrayList<>();
+
+	private final int partitionCount;
+
+	private final Function<Message<?>, Object> partitionKeyFunction;
+
+	private ThreadFactory threadFactory = new CustomizableThreadFactory("partition-thread-");
+
+	private boolean failover = true;
+
+	@Nullable
+	private LoadBalancingStrategy loadBalancingStrategy;
+
+	private ErrorHandler errorHandler;
+
+	private MessageHandlingTaskDecorator messageHandlingTaskDecorator = task -> task;
+
+	/**
+	 * Instantiate based on a provided number of partitions and function for partition key against
+	 * the message to dispatch.
+	 * @param partitionCount the number of partitions in this channel.
+	 * @param partitionKeyFunction the function to resolve a partition key against the message
+	 * to dispatch.
+	 */
+	public PartitionedDispatcher(int partitionCount, Function<Message<?>, Object> partitionKeyFunction) {
+		Assert.isTrue(partitionCount > 0, "'partitionCount' must be greater than 0");
+		Assert.notNull(partitionKeyFunction, "'partitionKeyFunction' must not be null");
+		this.partitionKeyFunction = partitionKeyFunction;
+		this.partitionCount = partitionCount;
+		populatedPartitions();
+	}
+
+	private void populatedPartitions() {
+		for (int i = 0; i < this.partitionCount; i++) {
+			this.partitions.put(i, newPartition());
+		}
+	}
+
+	private UnicastingDispatcher newPartition() {
+		ExecutorService executor = Executors.newSingleThreadExecutor(this.threadFactory);
+		this.executors.add(executor);
+		DelegateDispatcher delegateDispatcher =
+				new DelegateDispatcher(new ErrorHandlingTaskExecutor(executor, this.errorHandler));
+		delegateDispatcher.setFailover(this.failover);
+		delegateDispatcher.setLoadBalancingStrategy(this.loadBalancingStrategy);
+		delegateDispatcher.setMessageHandlingTaskDecorator(this.messageHandlingTaskDecorator);
+		return delegateDispatcher;
+	}
+
+
+	/**
+	 * Set a {@link ThreadFactory} for executors per partitions.
+	 * Defaults to the {@link CustomizableThreadFactory} based on a {@code partition-thread-} prefix.
+	 * @param threadFactory the {@link ThreadFactory} to use.
+	 */
+	public void setThreadFactory(ThreadFactory threadFactory) {
+		Assert.notNull(threadFactory, "'threadFactory' must not be null");
+		this.threadFactory = threadFactory;
+		populatedPartitions();
```

Comment:
Need to shut down and clear existing executors.

Perhaps call `shutdown()` first?

Reply:
Yeah... Nothing to shutdown I guess: those executors have nit been used yet.
But we indeed have to `this.executors.clear()` .

Reply:
We don't know that; they could call this after we started processing messages.

Reply:
And if not, it's a no-op; so why worry?

Reply:
Doh... We cannon populate them from the ctor: some other options are required for delegate dispatchers.
So, coming back to the `synchronized` solution 😄